pkg/server/sotw/v3/ads.go

```diff
@@ -0,0 +1,135 @@
+		nonce, err := sw.send(resp)
+		if err != nil {
+			return err
+		}
+
+		sw.watches.responders[resp.GetRequest().TypeUrl].nonce = nonce
+		return nil
+	}
+
+	processAllExcept := func(typeURL string) error {
+		for {
+			select {
+			// We watch the multiplexed ADS channel for incoming responses.
+			case res := <-sw.watches.responders[resource.AnyType].response:
+				if res.GetRequest().TypeUrl != typeURL {
+					if err := process(res); err != nil {
+						return err
+					}
+				}
+			default:
+				return nil
+			}
+		}
+	}
+
+	// This control loop strictly orders resources when running in ADS mode.
+	// It should be treated as a child process of the original process() loop
+	// and should return on close of stream or error. This will cause the
+	// cleanup routines in the parent process() loop to execute.
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		// We only watch the multiplexed channel since all values will come through from process.
+		case res := <-sw.watches.responders[resource.AnyType].response:
+			if err := process(res); err != nil {
+				return status.Errorf(codes.Unavailable, err.Error())
+			}
+		case req, ok := <-reqCh:
+			// Input stream ended or failed.
+			if !ok {
+				return nil
+			}
+
+			// Received an empty request over the request channel. Can't respond.
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Only first request is guaranteed to hold node info so if it's missing, reassign.
+			if req.Node != nil {
+				sw.node = req.Node
+			} else {
+				req.Node = sw.node
+			}
+
+			// Nonces can be reused across streams; we verify nonce only if nonce is not initialized.
+			nonce := req.GetResponseNonce()
+
+			// type URL is required for ADS but is implicit for xDS
+			if defaultTypeURL == resource.AnyType {
+				if req.TypeUrl == "" {
+					return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+				}
+			} else if req.TypeUrl == "" {
+				req.TypeUrl = defaultTypeURL
+			}
+
+			if s.callbacks != nil {
+				if err := s.callbacks.OnStreamRequest(sw.ID, req); err != nil {
+					return err
+				}
+			}
+
+			if lastResponse, ok := sw.lastDiscoveryResponses[req.TypeUrl]; ok {
+				if lastResponse.nonce == "" || lastResponse.nonce == nonce {
+					// Let's record Resource names that a client has received.
+					sw.streamState.SetKnownResourceNames(req.TypeUrl, lastResponse.resources)
+				}
+			}
+
+			typeURL := req.GetTypeUrl()
+			// Use the multiplexed channel for new watches.
+			responder := sw.watches.responders[resource.AnyType].response
+			if w, ok := sw.watches.responders[typeURL]; ok {
+				// We've found a pre-existing watch, lets check and update if needed.
+				// If these requirements aren't satisfied, leave an open watch.
+				if w.nonce == "" || w.nonce == nonce {
+					w.close()
+
+					// Only process if we have an existing watch otherwise go ahead and create.
+					if err := processAllExcept(typeURL); err != nil {
```

Comment:
@valerian-roche good call I'll drop a comment there but the original algorithm as outlined here:

https://github.com/envoyproxy/go-control-plane/pull/461

Reply:
It's been quite a long time since I've ported that logic over but if I remember correctly we don't actually drop the responses, they just get requeued and eventually are sent once the enforced order has been resolved. (CDS before EDS, LDS before RDS, etc...)

Reply:
From what I remember, here it is dequeuing all pre-existing responses in the channel. If of the type it seems to simply be discarded. I'm afraid of the impact if the version in the request is not changed, as the control-plane will consider there is no change and not send new resources

Reply:
Hm maybe we can write a test to cover that

Reply:
Quick update, the tests actually do cover partially sent snapshots!